### PR TITLE
fix: avoid extraneous upcast in backwards pass of dot_general

### LINKED ISSUE #34330: Avoid extraneous upcast in backwards pass of dot_general
### Description

In my forwards pass, I have a matmul with types `bf16 @ f32 -> bf16` (think activations times weights). In the backwards pass, I wish to compute the gradient with respect to the RHS (f32) input, so the backwards matmul should have the type `bf16 @ bf16 -> f32`.

However, JAX outputs inefficient code by first performing a `bf16 @ bf16 -> bf16` matmul and then upcasting the output to `f32`. I would like to have JAX directly output the matmul with the correct dtype without upcasts.

### Reproducer

```python
import jax
import jax.numpy as jnp

x = jnp.ones((3, 4), dtype=jnp.bfloat16)
w = jnp.ones((4, 5), dtype=jnp.float32)

def forward(w, x):
    y = jax.lax.dot_general(x, w,
                            dimension_numbers=((1, 0), (tuple(), tuple())),
                            preferred_element_type=jnp.bfloat16)
    return jnp.mean(y, dtype=jnp.bfloat16)

grad = jax.grad(forward)

print(jax.make_jaxpr(grad)(w, x))
```
Output:
```
{ lambda ; a:f32[4,5] b:bf16[3,4]. let
    c:bf16[3,5] = dot_general[
      dimension_numbers=(([1], [0]), ([], []))
      preferred_element_type=bfloat16
    ] b a
    d:bf16[] = reduce_sum[axes=(0, 1)] c
    _:bf16[] = div d 15.0:bf16[]
    e:bf16[] = div 1.0:bf16[] 15.0:bf16[]
    f:bf16[3,5] = broadcast_in_dim[
      broadcast_dimensions=()
      shape=(3, 5)
      sharding=None
    ] e
    g:bf16[5,4] = dot_general[
      dimension_numbers=(([0], [0]), ([], []))
      preferred_element_type=bfloat16     # <-- This is the wrong dtype!
    ] f b
    h:bf16[4,5] = transpose[permutation=(1, 0)] g
    i:f32[4,5] = convert_element_type[new_dtype=float32 weak_type=False] h    # <-- Unnecessary upcast
  in (i,) }
```
You can see here in the last 3 Jaxpr equations, there is a `bf16 @ bf16 -> bf16` matmul followed by an upcast to `f32`.

### Fix?

I think this could potentially be fixed by modifying the backwards rule for dot_general. For example, in this line of code:
https://github.com/jax-ml/jax/blob/fcbffe9cbf0c8e58bb36d77b7d42d561cbec4321/jax/_src/lax/lax.py#L5482-L5489

We could change it to something like:
```python
if x.aval.dtype in [float32, float16]:  # A list of generally supported types
  preferred_element_type = x.aval.dtype
dot_general_out = dot_general(g, y, dims, precision=precision,
                                preferred_element_type=preferred_element_type,
                                out_sharding=ds)
```

I'm not sure if there are any other considerations (e.g. backend-dependent behavior), but something like this worked for my use case.

I can send a PR if this solution seems satisfactory.

### System info (python version, jaxlib version, accelerator, etc.)

(This is a colab machine)

jax:    0.7.2
jaxlib: 0.7.2
numpy:  2.0.2
python: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='41d5f8880635', release='6.6.105+', version='#1 SMP Thu Oct  2 10:42:05 UTC 2025', machine='x86_64')

### PR DESCRIPTION
**Pull request by hrideymarwah15**
## Summary
Fix extraneous upcast in backwards pass of `dot_general` with mixed precision types.

## Problem
Link to issue: #34330

When using `dot_general` with mixed precision (e.g., bf16 @ f32 → bf16 via `preferred_element_type`), the backwards pass produces inefficient code. It performs a bf16 matmul followed by an unnecessary `convert_element_type` to f32.

**Before (buggy):**
```
g:bf16[5,4] = dot_general[...preferred_element_type=bfloat16]
h:bf16[4,5] = transpose[...] g
i:f32[4,5] = convert_element_type[new_dtype=float32] h  # Unnecessary!
```

**After (fixed):**
```
g:f32[5,4] = dot_general[...preferred_element_type=float32]  # Directly f32
h:f32[4,5] = transpose[...] g
# No conversion needed
```

## Solution
Use the target gradient dtype (`x.aval.dtype`) as `preferred_element_type` for the backward matmul in `_dot_general_transpose_lhs`, eliminating the need for the post-matmul conversion.

## Testing
- [x] Regression test added (`testDotGeneralBackwardNoExtraneousUpcast`)
- [x] All existing tests pass (2473 in lax_test.py, 22 dot-related autodiff tests)

Fixes #34330

### CODE REVIEW

**Comment by yashk2810** [comment]
Thanks for the contribution. We tried doing this thing before but it lead to issues internally. We'll try to fix this in a different way but for now we can't submit this PR so I am going to close it.
